### DISCUSSION TITLE
Switch to Symfony\Component\Uid for UUIDs

### DISCUSSION THREAD

**Discussion by binaryfire** [laravel/framework#47631] [Ideas]
According to the benchmarks `Symfony\Component\Uid` is more performant than `Ramsey\Uuid\Uuid`. The Symfony component is already installed for the UID helpers, so wouldn't it make sense to drop the Ramsey package altogether?

Links:
- https://dev.to/fredbouchery/performance-of-uuids-3i40 (see the updated benchmarks at the bottom of the article)
```
Ramsey => 2583 ms
ext-uuid => 1260 ms
polyfill-symfony => 4020 ms
polyfill-symfony-optim => 1145 ms (the current Symfony implementation)
custom => 642 ms
```

- https://jolicode.com/blog/uuid-generation-in-php

```
| benchPecl            | 4.546μs | 2.87x |
| benchSymfonyPolyfill | 1.583μs | 1.00x |
| benchJoliCodeFFI     | 5.416μs | 3.42x |
| benchRamsey          | 2.103μs | 1.33x |
```

**Comment by shengslogar** [comment]
Performance aside, Laravel references both Ramsey UUID and Symfony ULID helpers in the same files. Git Blame doesn't provide any insight, other than these changes were added by different contributors. At first glance, would think it'd make sense to refactor entirely to one library or the other.
https://github.com/laravel/framework/blob/fc8f819bef5b4843baadda393c51320ceeb41a70/src/Illuminate/Support/Str.php#L15-L17

**Comment by miken32** [comment]
Came here to ask the same thing years later, in light of some troubles getting updates of `ramsey/uuid`. I'm not an expert but I can't see any reason for having both libraries as dependencies.

**Comment by shaedrich** [comment]
⚠️ [`Symfony/uid`](https://github.com/Symfony/uid) does not support all UUID variants as [`ramsey/uuid`](https://github.com/ramsey/uuid) does (namely v2), therefore, using ramsey's implementation is necessary and not doing so would be a breaking change. (This was a mistake (or misassumption) I made in #53341 which I had to correct in #53368, see also #59661)

**Comment by miken32** [comment]
So it is just about being able to do `Str::isUuid($x, 2)`, or input validation with `uuid:2`, both of which were new features added in 12. I wonder what the impact would be of removing validation of a little-used UID format from the 1990s – they're not even part of an RFC.

**Reply by shaedrich** [reply]
I would ask the question how unexpected it would be for Laravel to not support this or, in other words, what we would have to add to the docs in order to make sure it's not unexpected and if we are willing to do this just because we want to throw out `ramsey/uuid` rather than `Symfony/uid`

**Reply by miken32** [reply]
I mean, it would just be a note in the docs, same as [Symfony has](https://symfony.com/doc/current/components/uid.html). 

I don't suggest this on a whim, "just because [I] want to throw out ramsey/uuid." I think being concerned about a package (which appears to be a one-man show) holding back updates [for weeks/months](https://github.com/ramsey/uuid/pull/638) is reasonable. Especially where an (almost) equivalent Symfony package exists, which we can reasonably expect to be well maintained and promptly updated.

**Reply by shaedrich** [reply]
Would be two notes (one for `Str` and one for validation)

Have there been attempts to fork `ramsey/uuid`? If not, would it be worth to provide a small wrapper in Laravel itself (`Illuminate\Support\Uuid`) so we can have the best out of both worlds with minimal overhead should people need v2?

**Reply by miken32** [reply]
IMO if someone has the ability to write that wrapper for Laravel, it would be better to submit upstream to the Symfony component. Unless they have a reason they aren't supporting it beyond a lack of use.

But as I type that I realize I'm thinking about generating UUIDs, which Laravel doesn't do except for versions 4 and 7.  Validation is the only question here, and as far as I understand it, a valid UUID v2 should also be a valid UUID v1 (with the exception of the version digit.)

But everything I know about UUID v2 comes from [this page](https://playfulprogramming.com/posts/what-happened-to-uuid-v2/) I read 10 minutes ago, so I may not know what I'm talking about.

**Reply by shaedrich** [reply]
> it would be better to submit upstream to the Symfony component. Unless they have a reason they aren't supporting it beyond a lack of use.

Their lack of it might be telling already but I might be wrong

> [this page](https://playfulprogramming.com/posts/what-happened-to-uuid-v2/)

This was super interesting—thanks for the link 👍🏻 💡 

Would be worth asking if Laravel sees himself as someone preventing someone else from making a bad choice 🤔

**Reply by miken32** [reply]
Well, it is described as "opinionated"
